File: dataPreprocessing_v2/KmlWriter.py
import logging
import GPSCoordinate
from PSWF import PSWF


class KmlWriter:
    def __init__(self, bfn):
        self.base_filename = bfn
        self.kml_filename = bfn + ".kml"

        self.t2 = "  "
        self.t4 = "    "
        self.t6 = "      "
        self.t8 = "        "
        self.t10 = "          "

        self.pal2 = "http://maps.google.com/mapfiles/kml/pal2/"
        self.pal3 = "http://maps.google.com/mapfiles/kml/pal3/"
        self.pal4 = "http://maps.google.com/mapfiles/kml/pal4/"
        self.pal5 = "http://maps.google.com/mapfiles/kml/pal5/"

        self.header_written = False
        self.footer_written = False

        self.buffered_writer = open(self.kml_filename, "w")

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.buffered_writer.close()

    def write_kml_name_description(self, n, d):
        if self.buffered_writer:
            self.buffered_writer.write(f"{self.t4}<name>{n}</name>\n")
            self.buffered_writer.write(f"{self.t4}<open>1</open>\n")
            self.buffered_writer.write(f"{self.t4}<description>{d}</description>\n")
            return True
        return False

    def write_folder_header(self, name, descr):
        if self.buffered_writer:
            self.buffered_writer.write(f"{self.t4}<Folder>\n")
            self.buffered_writer.write(f"{self.t6}<name>{name}</name>\n")
            self.buffered_writer.write(f"{self.t6}<description>{descr}</description>\n")
            return True
        return False

    def write_folder_footer(self):
        if self.buffered_writer:
            self.buffered_writer.write(f"{self.t4}</Folder>\n")
            return True
        return False

    def write_kml_header(self):
        if self.buffered_writer and not self.header_written:
            self.buffered_writer.write("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n")
            self.buffered_writer.write("<kml xmlns=\"http://www.opengis.net/kml/2.2\">\n")
            self.buffered_writer.write(f"{self.t2}<Document>\n")
            self.header_written = True
            return True
        return False

    def write_kml_footer(self):
        if self.buffered_writer and self.header_written and not self.footer_written:
            self.buffered_writer.write(f"{self.t2}</Document>\n")
            self.buffered_writer.write("</kml>\n")
            self.buffered_writer.close()
            self.footer_written = True
            return True
        return False

    def write_style_ids(self) -> bool:
        return (
                self.write_gps_sid()
                and self.write_pswf_sid()
                and self.write_pres_seg_sid()
                and self.write_veh_gps_path_sid()
        )

    def write_line_style(self, name: str, color: str, width: str) -> bool:
        if self.buffered_writer:
            try:
                self.buffered_writer.write(self.t4 + f'<Style id="{name}">\n')
                self.buffered_writer.write(self.t6 + "<LineStyle>\n")
                self.buffered_writer.write(self.t8 + f"<color>{color}</color>\n")
                self.buffered_writer.write(self.t8 + f"<width>{width}</width>\n")
                self.buffered_writer.write(self.t6 + "</LineStyle>\n")
                self.buffered_writer.write(self.t4 + "</Style>\n")
                return True
            except IOError as ex:
                logging.exception("Exception in KmlWriter.write_line_style", exc_info=ex)
        return False

    def write_icon_style(self, name: str, url: str) -> bool:
        if self.buffered_writer:
            try:
                self.buffered_writer.write(self.t4 + f'<Style id="{name}">\n')
                self.buffered_writer.write(self.t6 + "<IconStyle>\n")
                self.buffered_writer.write(self.t8 + "<Icon>\n")
                self.buffered_writer.write(self.t10 + f"<href>{url}</href>\n")
                self.buffered_writer.write(self.t8 + "</Icon>\n")
                self.buffered_writer.write(self.t6 + "</IconStyle>\n")
                self.buffered_writer.write(self.t4 + "</Style>\n")
                return True
            except IOError as ex:
                logging.exception("Exception in KmlWriter.write_icon_style", exc_info=ex)
        return False

    def write_line_string_header(self, extrude: str, altitude: str, tessellate: str) -> bool:
        if self.buffered_writer:
            try:
                self.buffered_writer.write(self.t4 + "<Placemark>\n")
                self.buffered_writer.write(self.t4 + "<styleUrl>#VehGpsPathStyle</styleUrl>\n")
                self.buffered_writer.write(self.t4 + "<LineString>\n")
                self.buffered_writer.write(self.t6 + "<extrude>" + extrude + "</extrude>\n")
                self.buffered_writer.write(self.t6 + "<altitudeMode>" + altitude + "</altitudeMode>\n")
                self.buffered_writer.write(self.t6 + "<tessellate>" + tessellate + "</tessellate>\n")
                self.buffered_writer.write(self.t8 + "<coordinates>\n")

                return True
            except IOError as ex:
                logging.getLogger('KmlWriter').exception(ex)
                return False
        return False

    def write_line_string_coordinate(self, lon: float, lat: float) -> bool:
        if self.buffered_writer:
            try:
                self.buffered_writer.write(self.t8 + str(lon) + "," + str(lat) + ",1\n")

                return True
            except IOError as ex:
                logging.getLogger('KmlWriter').exception(ex)
                return False
        return False

    def write_line_string_footer(self) -> bool:
        if self.buffered_writer:
            try:
                self.buffered_writer.write(self.t8 + "</coordinates>\n")
                self.buffered_writer.write(self.t6 + "</LineString>\n")
                self.buffered_writer.write(self.t6 + "</Placemark>\n")

                return True
            except IOError as ex:
                logging.getLogger('KmlWriter').exception(ex)
                return False
        return False

    def write_placemark(self, name: str, descr: str, icon_style: str, gps: GPSCoordinate) -> bool:
        if self.buffered_writer:
            try:
                self.buffered_writer.write(self.t4 + "<Placemark>\n")
                self.buffered_writer.write(self.t6 + "<name>" + name + "</name>\n")
                self.buffered_writer.write(self.t6 + "<description>" + descr + "</description>\n")
                self.buffered_writer.write(self.t6 + "<styleUrl>#" + icon_style + "</styleUrl>\n")
                self.buffered_writer.write(self.t6 + "<Point>\n")
                self.buffered_writer.write(
                    self.t8 + "<coordinates>" + str(gps.longitude) + "," + str(gps.latitude) + ",0</coordinates>\n")
                self.buffered_writer.write(self.t6 + "</Point>\n")
                self.buffered_writer.write(self.t4 + "</Placemark>\n")

                return True
            except IOError as ex:
                logging.getLogger('KmlWriter').exception(ex)
                return False
        return False

    def write_gps_placemark(self, name: str, descr: str, gps: GPSCoordinate) -> bool:
        return self.write_placemark(name, descr, "GPSIconStyle", gps)

    def write_pres_seg_placemark(self, name: str, descr: str, gps: GPSCoordinate) -> bool:
        return self.write_placemark(name, descr, "PresentSegmentIconStyle", gps)

    def write_pswf_placemark(self, pswf: PSWF, name: str, descr: str, gps: GPSCoordinate) -> bool:
        return self.write_placemark(name, descr, pswf.name + "IconStyle", gps)

    def write_gps_sid(self) -> bool:
        return self.write_icon_style("GPSIconStyle", self.pal4 + "icon57.png")

    def write_pres_seg_sid(self) -> bool:
        return self.write_icon_style("PresentSegmentIconStyle", self.pal4 + "icon56.png")

    def write_pswf_sid(self) -> bool:
        return (
                self.write_icon_style("ParkenIconStyle", self.pal5 + "icon47.png")
                and self.write_icon_style("StandfunktionenIconStyle", self.pal5 + "icon26.png")
                and self.write_icon_style("WohnenIconStyle", self.pal5 + "icon30.png")
                and self.write_icon_style("PruefenAnDiIconStyle", self.pal5 + "icon47.png")
                and self.write_icon_style("FahrenIconStyle", self.pal5 + "icon61.png")
                and self.write_icon_style("InvalidIconStyle", self.pal5 + "icon40.png"))

    def write_veh_gps_path_sid(self) -> bool:
        return self.write_line_style("VehGpsPathStyle", "ffffffff", "2")

File: 2023.04.28_asc2kml/src/asc2kml/KmlWriter.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package asc2kml;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author Q262579
 */
public class KmlWriter {
    
    public KmlWriter(String bfn) {
        baseFilename = bfn;
        kmlFilename = baseFilename + kmlExtension;                
        
        try {
            fileWriter = new FileWriter(kmlFilename);
        } catch (IOException ex) {
            Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
        }
        
        if (fileWriter != null) {
            bW = new BufferedWriter(fileWriter);
        }
        
    }
    
    public boolean writeKmlNameDescription(String n, String d) {
        if (bW != null) {
            try {
                bW.write(t4 + "<name>" + n + "</name>"); bW.newLine();
                bW.write(t4 + "<open>1</open>"); bW.newLine();
                bW.write(t4 + "<description>" + d + "</description>"); bW.newLine();
                return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    public boolean writeFolderHeader(String name, String descr, boolean visibility, boolean open) {
        if (bW != null) {
            try {                                
                bW.write(t4 + "<Folder>"); bW.newLine();
                bW.write(t6 + "<name>" + name + "</name>"); bW.newLine();
                bW.write(t6 + "<description>" + descr + "</description>"); bW.newLine();
                bW.write(t6 + "<visibility>" + (visibility ? "1" : "0") + "</visibility>"); bW.newLine();
                bW.write(t6 + "<open>" + (open ? "1" : "0") + "</open>"); bW.newLine();
                
                return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    public boolean writeFolderFooter() {
        if (bW != null) {
            try {                                
                bW.write(t4 + "</Folder>"); bW.newLine();                
                
                return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    public boolean writeKmlHeader() {
        if (bW != null) {
            if (!headerWritten) {
                try {
                    bW.write("<?xml version=\"1.0\" encoding=\"UTF-8\"?>"); bW.newLine();
                    bW.write("<kml xmlns=\"http://www.opengis.net/kml/2.2\">"); bW.newLine();
                    bW.write(t2 + "<Document>"); bW.newLine();                    
                    headerWritten = true;
                    return true;
                } catch (IOException ex) {
                    Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                    return false;
                }
            }
        }
        return false;
    }
    
    public boolean writeKmlFooter() {
        if (bW != null) {
            if (headerWritten && !footerWritten) {                
                try {
                    bW.write(t2 + "</Document>"); bW.newLine();
                    bW.write("</kml>"); bW.newLine();
                    
                    bW.close();
                    
                    footerWritten = true;
                    return true;
                } catch (IOException ex) {
                    Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                    return false;
                }
            }
        }
        return false;
    }
    
    public boolean writeStyleIDs() {
        if (writeGPS_SID() &&
            writePSWF_SID() &&
            writeEntryExit_SID() &&
            writeExitHistory_SID() &&
            writePRES_SEG_SID() &&
            writeVehGpsPath_SID() &&
            writeVehProp_SID()
        ) {
            return true;
        }
        return false;
    }    
    
    private boolean writeLineStyle(String name, String color, String width) {
        if (bW != null) {
            try {                                
                    bW.write(t4 + "<Style id=\"" + name + "\">"); bW.newLine();
                    bW.write(t6 + "<LineStyle>"); bW.newLine();
                    bW.write(t8 + "<color>" + color + "</color>"); bW.newLine();
                    bW.write(t8 + "<width>" + width + "</width>"); bW.newLine();
                    bW.write(t6 + "</LineStyle>"); bW.newLine();
                    bW.write(t4 + "</Style>"); bW.newLine();                    
                
                    return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    private boolean writeIconStyle(String name, String url) {
        if (bW != null) {
            try {                                
                    bW.write(t4 + "<Style id=\"" + name + "\">"); bW.newLine();
                    bW.write(t6 + "<IconStyle>"); bW.newLine();
                    bW.write(t8 + "<Icon>"); bW.newLine();
                    bW.write(t10 + "<href>" + url + "</href>"); bW.newLine();
                    bW.write(t8 + "</Icon>"); bW.newLine();
                    bW.write(t6 + "</IconStyle>"); bW.newLine();
                    bW.write(t4 + "</Style>"); bW.newLine();
                
                    return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    public boolean writeLineStringHeader(String extrude, String altitude, String tessallate) {
        if (bW != null) {
            try {
                bW.write(t4 + "<Placemark>"); bW.newLine();
                bW.write(t4 + "<styleUrl>#VehGpsPathStyle</styleUrl>"); bW.newLine();
                bW.write(t4 + "<LineString>"); bW.newLine();
                bW.write(t6 + "<extrude>" + extrude + "</extrude>"); bW.newLine();
                bW.write(t6 + "<altitudeMode>" + altitude + "</altitudeMode>"); bW.newLine();     
                bW.write(t6 + "<tessallate>" + tessallate + "</tessallate>"); bW.newLine();                
                bW.write(t8 + "<coordinates>"); bW.newLine();                
                
                return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    public boolean writeLineStringCoordinate(double lon, double lat) {
        if (bW != null) {
            try {                                
                bW.write(t8 + lon + "," + lat + ",1"); bW.newLine();                
                
                return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    public boolean writeLineStringFooter() {
        if (bW != null) {
            try {                                
                bW.write(t8 + "</coordinates>"); bW.newLine();
                bW.write(t6 + "</LineString>"); bW.newLine();
                bW.write(t6 + "</Placemark>"); bW.newLine();
                
                return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    public boolean writePlacemark(String name, String descr, String iconStyle, GPSCoordinate gps) {
        if (bW != null) {
            try {                                
                bW.write(t4 + "<Placemark>"); bW.newLine();
                bW.write(t6 + "<name>" + name + "</name>"); bW.newLine();
                bW.write(t6 + "<description>" + descr + "</description>"); bW.newLine();     
                bW.write(t6 + "<styleUrl>#" + iconStyle + "</styleUrl>"); bW.newLine();
                //bW.write(t6 + "<font face=verdana>Verdana</font>"); bW.newLine();
                bW.write(t6 + "<Point>"); bW.newLine();
                bW.write(t8 + "<coordinates>" + gps.longitude + "," + gps.latitude + ",0</coordinates>"); bW.newLine();
                bW.write(t6 + "</Point>"); bW.newLine();
                bW.write(t4 + "</Placemark>"); bW.newLine();
                
                return true;
            } catch (IOException ex) {
                Logger.getLogger(KmlWriter.class.getName()).log(Level.SEVERE, null, ex);
                return false;
            }                        
        }
        return false;
    }
    
    public boolean writeGPSPlacemark(String name, String descr, GPSCoordinate gps) {
        if (writePlacemark(name, descr, "GPSIconStyle", gps)) {
            return true;
        }
        return false;
    }
    
    public boolean writePRES_SEGPlacemark(String name, String descr, GPSCoordinate gps) {
        if (writePlacemark(name, descr, "PresentSegmentIconStyle", gps)) {
            return true;
        }
        return false;
    }
    
    public boolean writePSWFPlacemark(String name, String descr, PSWF pswf, GPSCoordinate gps) {
        if (writePlacemark(name, descr, pswf.label + "IconStyle", gps)) {
            return true;
        }
        return false;        
    }
    
    public boolean writeEntryExitPlacemark(String name, String descr, String type, GPSCoordinate gps) {
        if (writePlacemark(name, descr, type + "IconStyle", gps)) {
            return true;
        }
        return false;        
    }
    
    public boolean writeExitHistoryPlacemark(String name, String descr, GPSCoordinate gps) {
        if (writePlacemark(name, descr, "ExitHistoryIconStyle", gps)) {
            return true;
        }
        return false;        
    }
    
    public boolean writeLidsFlapsPlacemark(String name, String descr, GPSCoordinate gps) {
        if (writePlacemark(name, descr, "VehiclePropertyIconStyle", gps)) {
            return true;
        }
        return false;        
    }
    
    private boolean writeGPS_SID() {
        if (writeIconStyle("GPSIconStyle", pal4 + "icon57.png"))
            return true;
        return false;
    }
    
    private boolean writePRES_SEG_SID() {
        if (writeIconStyle("PresentSegmentIconStyle", pal4 + "icon56.png"))
            return true;
        return false;
    }
    
    private boolean writePSWF_SID() {
        if (writeIconStyle("ParkenIconStyle", pal5 + "icon47.png") &&
            writeIconStyle("StandfunktionenIconStyle", pal5 + "icon26.png") &&
            writeIconStyle("WohnenIconStyle", pal5 + "icon30.png") &&
            writeIconStyle("PruefenAnDiIconStyle", pal5 + "icon47.png") &&
            writeIconStyle("FahrenIconStyle", pal5 + "icon61.png") &&
            writeIconStyle("InvalidIconStyle", pal5 + "icon40.png")
           )
            return true;
        return false;
    }        
    
    private boolean writeEntryExit_SID() {
        if (writeIconStyle("EntryIconStyle", local + "icon54(2).png") &&            
            writeIconStyle("ExitIconStyle", local + "icon7(2).png")
           )
            return true;
        return false;
    }
    
    private boolean writeExitHistory_SID() {
        if (writeIconStyle("ExitHistoryIconStyle", pal4 + "icon8.png"))
            return true;
        return false;
    }
    
    private boolean writeVehGpsPath_SID() {
        if (writeLineStyle("VehGpsPathStyle", "ffffffff", "2"))
            return true;
        return false;
    }
    
    private boolean writeVehProp_SID() {
        if (writeIconStyle("VehiclePropertyIconStyle", pal3 + "icon62.png"))
            return true;
        return false;
    }
    
    private String baseFilename = null;
    private String kmlFilename = null;
    private final String kmlExtension = ".kml";
    
    private final String t2 = "  ";
    private final String t4 = "    ";
    private final String t6 = "      ";
    private final String t8 = "        ";
    private final String t10 = "          ";
    
    private final String local = "C:\\Program Files\\GoogleEarthIcons\\";
    private final String pal2 = "http://maps.google.com/mapfiles/kml/pal2/";
    private final String pal3 = "http://maps.google.com/mapfiles/kml/pal3/";
    private final String pal4 = "http://maps.google.com/mapfiles/kml/pal4/";
    private final String pal5 = "http://maps.google.com/mapfiles/kml/pal5/";
    
    private boolean headerWritten = false;
    private boolean footerWritten = false;
    
    BufferedWriter bW = null;
    FileWriter fileWriter = null;
}
